### PR TITLE
Fixed #30023 -- Prevented SQLite schema alterations while foreign key checks are enabled.

### PR DESCRIPTION
**Pull request by charettes**
Prior to this change foreign key constraint references could be left pointing at tables dropped during operations simulating unsupported table alterations  because of an unexpected failure to disable foreign key constraint checks.
    
SQLite3 does not allow disabling such checks while in a transaction so they must be disabled beforehand.

### CODE REVIEW

File: django/db/backends/sqlite3/base.py [hunk 1]
```diff
@@ -247,11 +247,10 @@ def _set_autocommit(self, autocommit):
...
 
     def disable_constraint_checking(self):
-        if self.in_atomic_block:
-            # sqlite3 cannot disable constraint checking inside a transaction.
-            return False
-        self.cursor().execute('PRAGMA foreign_keys = OFF')
-        return True
+        with self.cursor() as cursor:
+            cursor.execute('PRAGMA foreign_keys = OFF')
+            enabled = cursor.execute('PRAGMA foreign_keys').fetchone()[0]
```
Comment (carltongibson) [inline django/db/backends/sqlite3/base.py]: Maybe a comment here? 

`...foreign key constraints cannot be disabled within open transaction...` or similar — this is Dark Arts here, any signposts for the future handy. 🙂

File: tests/backends/sqlite/tests.py [hunk 2]
```diff
@@ -82,6 +66,52 @@ class SchemaTests(TransactionTestCase):
 
     available_apps = ['backends']
 
+    def test_autoincrement(self):
```
Comment (carltongibson) [inline tests/backends/sqlite/tests.py]: Should moving this test go in a separate commit?

File: tests/indexes/tests.py [hunk 3]
```diff
@@ -25,12 +25,12 @@ def test_index_name_hash(self):
         """
         Index names should be deterministic.
         """
-        with connection.schema_editor() as editor:
```
Comment (carltongibson) [inline tests/indexes/tests.py]: I don't follow _Why_ these changes here?

File: tests/model_options/test_tablespaces.py [hunk 4]
```diff
@@ -21,7 +23,9 @@ def sql_for_index(model):
 # We can't test the DEFAULT_TABLESPACE and DEFAULT_INDEX_TABLESPACE settings
 # because they're evaluated when the model class is defined. As a consequence,
 # @override_settings doesn't work, and the tests depend
-class TablespacesTests(TestCase):
+class TablespacesTests(TransactionTestCase):
```
Comment (carltongibson) [inline tests/model_options/test_tablespaces.py]: What's at stake with the change to `TransactionTestCase` here? Is it just an optimisation? (If so, separate commit?)

File: tests/indexes/tests.py (hunk 3 above)
Comment (charettes) [reply tests/indexes/tests.py]: @carltongibson they `__enter__` a `SchemaEditor` within a transaction (because of the `TestCase` transaction wrapping) which on SQLite as that would prevent proper constraint disabling. There's no need to `__enter__` these schema editor instances at all to call this method as the following `test_index_together` and `test_index_together_single_list` tests demonstrate.

File: tests/model_options/test_tablespaces.py (hunk 4 above)
Comment (charettes) [reply tests/model_options/test_tablespaces.py]: @carltongibson the `sql_for_table` calls which is defined above have to `SchemaEditor.__enter__` which is now prohibited in a transaction. Using `TransactionTestCase` makes sure we don't wrap tests in a transaction.

File: django/db/backends/sqlite3/base.py (hunk 1 above)
Comment (charettes) [reply django/db/backends/sqlite3/base.py]: Makes sense.

**Comment by carltongibson** [comment]
Hey @charettes. Thanks for helping my understanding. 

> ... in a TestCase which wraps tests in a transaction.

Explains everything. 

In light of that, I'm happy with this as is. (Bar the comment in `disable_constraint_checking(), which, if you agree, I think would be good. I can add that if you like.)

**Comment by carltongibson** [comment]
Hey @charettes, just so you know, I have to pause now. I will merge this and the related PRs later on. Thank you for your super effort on these bugs!

**Comment by charettes** [comment]
No problem @carltongibson, I should find time to add the comment by then!

**Comment by carltongibson** [comment]
No issues. If you haven't done it I'll add one. (It'll give Tim something to adjust later 😀)

**Comment by charettes** [comment]
Just added a comment.

File: docs/releases/2.0.10.txt [hunk 5]
```diff
@@ -16,3 +16,6 @@ Bugfixes
 * Fixed a schema corruption issue on SQLite 3.26+. You might have to drop and
   rebuild your SQLite database if you applied a migration while using an older
   version of Django with SQLite 3.26 or later (:ticket:`29182`).
+
+* Fixed a schema corruption issue on SQLite when wrapping schema alterations
```
Comment (timgraham) [inline docs/releases/2.0.10.txt]: Prevented SQLite schema alterations while foreign key checks are enabled to avoid the possibility of schema corruption.

**Comment by carltongibson** [comment]
Hey @charettes. Just checking: this is intended to be back ported to 2.0.x yes?

**Comment by charettes** [comment]
@carltongibson right, I forgot about it but I'll take care of it now.

**Comment by charettes** [comment]
@carltongibson done!

**Comment by carltongibson** [comment]
Well done Simon. Thank you again for all your efforts on this group of issues!

File: tests/backends/sqlite/tests.py [hunk 6]
```diff
@@ -82,6 +66,52 @@ class SchemaTests(TransactionTestCase):
...
+    def test_disable_constraint_checking_failure_disallowed(self):
+        """
+        SQLite schema editor is not usable within an outer transaction if
+        foreign key constraint checks are not disabled beforehand.
+        """
+        msg = (
+            'SQLite schema editor cannot be used while foreign key '
+            'constraint checks are enabled. Make sure to disable them '
+            'before entering a transaction.atomic() context because '
+            'SQLite3 does not support disabling them in the middle of '
```
Comment (timgraham) [inline tests/backends/sqlite/tests.py]: The same comment about SQLite3 -> SQLite applies here and in the docstring below. Maybe you'd like to audit the rest of the Django repo as there are a few other places that might be worth changing. I'm unsure if there's some meaningful distinction between the two terms.
